cmd/lotus-shed/miner.go

```diff
@@ -354,3 +361,220 @@ var minerUnpackInfoCmd = &cli.Command{
+			return xerrors.Errorf("getting api: %w", err)
+		}
+		defer acloser()
+
+		ctx := lcli.ReqContext(cctx)
+
+		maddr, err := address.NewFromString(cctx.String("actor"))
+		if err != nil {
+			return xerrors.Errorf("getting actor address: %w", err)
+		}
+
+		minfo, err := api.StateMinerInfo(ctx, maddr, types.EmptyTSK)
+		if err != nil {
+			return xerrors.Errorf("getting mienr info: %w", err)
+		}
+
+		deadline, err := api.StateMinerProvingDeadline(ctx, maddr, types.EmptyTSK)
+		if err != nil {
+			return xerrors.Errorf("getting deadline: %w", err)
+		}
+
+		partitionIndices := cctx.Int64Slice("partitions")
+		if len(partitionIndices) <= 0 {
+			return fmt.Errorf("must include at least one partition to compact")
+		}
+
+		chainHead, err := api.ChainHead(ctx)
+		if err != nil {
+			return xerrors.Errorf("getting chain head: %w", err)
+		}
+
+		checkRand, err := api.StateGetRandomnessFromTickets(ctx, crypto.DomainSeparationTag_PoStChainCommit, deadline.Challenge, nil, chainHead.Key())
+		if err != nil {
+			return xerrors.Errorf("getting randomness: %w", err)
+		}
+
+		proofSize, err := minfo.WindowPoStProofType.ProofSize()
+		if err != nil {
+			return xerrors.Errorf("getting proof size: %w", err)
+		}
+
+		var partitions []miner8.PoStPartition
+
+		emptyProof := []proof.PoStProof{{
+			PoStProof:  minfo.WindowPoStProofType,
+			ProofBytes: make([]byte, proofSize)}}
+
+		for _, partition := range partitionIndices {
+			newPartition := miner8.PoStPartition{
+				Index:   uint64(partition),
+				Skipped: bitfield.New(),
+			}
+			partitions = append(partitions, newPartition)
+		}
+
+		params := miner8.SubmitWindowedPoStParams{
+			Deadline:         deadline.Index,
+			Partitions:       partitions,
+			Proofs:           emptyProof,
+			ChainCommitEpoch: deadline.Challenge,
+			ChainCommitRand:  checkRand,
+		}
+
+		sp, err := actors.SerializeParams(&params)
+		if err != nil {
+			return xerrors.Errorf("serializing params: %w", err)
+		}
+
+		fmt.Printf("submitting bad PoST for %d paritions\n", len(partitionIndices))
+		smsg, err := api.MpoolPushMessage(ctx, &types.Message{
+			From:   minfo.Worker,
+			To:     maddr,
+			Method: builtin.MethodsMiner.SubmitWindowedPoSt,
+			Value:  big.Zero(),
+			Params: sp,
+		}, nil)
+		if err != nil {
+			return xerrors.Errorf("mpool push: %w", err)
+		}
+
+		fmt.Println("Message CID:", smsg.Cid())
+
+		return nil
+	},
+}
+
+var generateAndSendConsensusFaultCmd = &cli.Command{
+	Name:        "generate-and-send-consensus-fault",
+	Usage:       "Provided a block CID mined by the miner, will create another block at the same height, and send both block headers to generate a consensus fault.",
+	Description: `Note: This is meant for testing purposes and should NOT be used on mainnet or you will be slashed`,
+	Flags: []cli.Flag{
+		&cli.BoolFlag{
+			Name:  "really-do-it",
+			Usage: "Actually send transaction performing the action",
+			Value: false,
+		},
+		&cli.StringFlag{
+			Name:  "actor",
+			Usage: "TODO",
```

Comment:
You don't really need this -- you can figure out the miner from the block being inputted.